swarm/src/lib.rs

```diff
@@ -511,7 +507,11 @@ where
                 Ok(address) => {
                     let transport_dial_opts = {
                         let transport_port_use = dial_opts.port_use().unwrap_or_else(|| {
-                            if self.listener_presence.contains(&address) {
+                            let is_listening_on_same_transport = self
+                                .listeners()
+                                .map(clean_multiaddr)
+                                .any(|e| e == clean_multiaddr(&address));
```

Comment:
Can you extract this into a utility function:

```rust
fn any_with_same_transport(haystack: impl Iterator, needle: &Multiaddr) -> bool { }
```

We can then easily write unit tests against it.